### ISSUE TITLE
get_rejection_threshold stops optimizing threshold as soon as one threshold leads to rejection of all channels in at least one fold

### ISSUE THREAD

**Issue by thht** [autoreject/autoreject#178]
hi,
i dug a bit into the global threshold finding function `get_rejection_threshold`. i do not know whether this is actually the desired behavior or constitutes an issue. but it looks strange enough for me to point it out:

1. if the current rejection threshold leads to all channels being rejected, line 172 in `autoreject.py` will return `nan`
  `self.mean_ = _slicemean(X, keep, axis=0)`
2. if this happens, the if clause at line 138 in the same file will evaluate to `true`, thus the scoring function will return `-Inf`.
3. line 251 in the same file:
  `obj = -np.mean(cross_val_score(est, X, cv=cv))`
  will evaluate to `-Inf` if this happens in at least one of the cross validation folds.
4. if this happens during the bayes optimization phase, line 60 in `bayesopt.py` will evaluate to `false`, which in turn leads to `new_x` not getting updated so the bayes optimization gets stuck with the previous `best_x`.

again, maybe this is exactly how this is supposed to work. but then i wonder why the `bayes_opt` does not terminate in such a case.

i would like to add that i tried this with a couple of Neuromag data files with decent data quality of different subjects and it always happened for both magnetometers and gradiometers.

best,
thomas

**Comment by jasmainak** [comment]
humm interesting, that sounds like a bug to me. I think I verified before merging the [pull request to speed up using bayesian optimization](https://github.com/autoreject/autoreject/pull/77) that it gave the same result as a grid search.

do you have a suggested fix for this?

**Comment by thht** [comment]
well, i have no idea what the intended behavior in this case would be. would the score still be valid if X out of N would be `-Inf`? What would X and N be?

If the score (and thus the `new_x` chosen in `bayes_opt`) is deemed to be invalid, how can we tell the algorithm to choose some different `new_x` to try?

**Comment by jasmainak** [comment]
ummm ... I think the simplest might be to leave out X which give `-Inf` for any fold since they correspond to a very low threshold (= rejecting too many trials). This should be done before starting the optimization. @agramfort do you have an opinion here?

**Comment by agramfort** [comment]
yes we could give bounds in optim grid to avoid this case before staring
the opt.

>

**Comment by jasmainak** [comment]
@thht can you share a dataset on which we can reproduce this?

**Comment by sappelhoff** [comment]
I pinged him in an email, we may have a chance for him to come back later in august :-)

**Comment by thht** [comment]
hi,
sorry, this would need some work so i can remember what dataset it was etc... i am still on paternity leave, as @sappelhoff mentioned, until mid-august. i might be able to take a look at it in the second half of july because the little one will slowly start daycare then.... but until then, i do not have the time for it. sorry again!

**Comment by jasmainak** [comment]
okay! no rush and thank you for this :)

**Comment by thht** [comment]
@jasmainak i just wrote you an email with a link to a dataset that replicates the issue for me, also with the fix merged.

if you do something along these lines:
```
raw_data = mne.io.Raw('data/autoreject_issue.fif')

art_events = mne.make_fixed_length_events(raw_data, duration=1)

epochs_for_art = mne.Epochs(raw_data, events=art_events,
                            tmin=0, tmax=1,
                            proj=True, reject=None, preload=True,
                            baseline=None, detrend=0)

rejection_dict2 = autoreject.get_rejection_threshold(epochs_for_art,
                                                     verbose='tqdm')
```
you are going to see that in `bayesopt.py` in line 60,`f_new` is always `Inf` so the value is not updated#...
